semtypes/src/main/java/io/ballerina/types/PredefinedTypeEnv.java

```diff
@@ -0,0 +1,345 @@
+/*
+ *  Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.types;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.ballerina.types.Core.union;
+import static io.ballerina.types.PredefinedType.CELL_SEMTYPE_INNER_MAPPING;
+import static io.ballerina.types.PredefinedType.CELL_SEMTYPE_INNER_MAPPING_RO;
+import static io.ballerina.types.PredefinedType.CELL_SEMTYPE_INNER_RO;
+import static io.ballerina.types.PredefinedType.CELL_SEMTYPE_UNDEF;
+import static io.ballerina.types.PredefinedType.CELL_SEMTYPE_VAL;
+import static io.ballerina.types.PredefinedType.INNER;
+import static io.ballerina.types.PredefinedType.INNER_READONLY;
+import static io.ballerina.types.PredefinedType.MAPPING;
+import static io.ballerina.types.PredefinedType.MAPPING_RO;
+import static io.ballerina.types.PredefinedType.NEVER;
+import static io.ballerina.types.PredefinedType.UNDEF;
+import static io.ballerina.types.PredefinedType.VAL;
+import static io.ballerina.types.TypeAtom.createTypeAtom;
+
+/**
+ * This is a utility class used to create various type atoms that needs to be initialized without an environment and
+ * common to all environments. When we construct an {@code Env}, we can use {@code initializeEnv} to populate it with
+ * those atoms.
+ * NOTE: While this class lazy initialize all the atoms technically {@code PredefinedType} will cause it initialize
+ * all the atoms currently.
+ * @since 2201.10.0
+ */
+public final class PredefinedTypeEnv {
+
+    private PredefinedTypeEnv() {
+    }
+
+    private static final PredefinedTypeEnv INSTANCE = new PredefinedTypeEnv();
+
+    public static PredefinedTypeEnv getInstance() {
+        return INSTANCE;
+    }
+
+    private final List<InitializedTypeAtom<CellAtomicType>> initializedCellAtoms = new ArrayList<>();
+    private final List<InitializedTypeAtom<ListAtomicType>> initializedListAtoms = new ArrayList<>();
+    private final List<ListAtomicType> initializedRecListAtoms = new ArrayList<>();
+    private final List<MappingAtomicType> initializedRecMappingAtoms = new ArrayList<>();
+    // 0 is reserved for BDD_REC_ATOM_READONLY
+    private AtomicInteger nextAtomIndex = new AtomicInteger(1);
```

Comment:
At L309; `initializedCellAtoms.forEach(each -> env.cellAtom(each.atomicType()));` when we call `env.cellAtom()` for the first time. it will have the index 0 I guess. Since the env.atomTable is empty at that time.

So shouldn't we start with index 0. Should we also call this nextTypeAtomIndex? Becuase this is need for only the typeAtoms. Not the recAtoms?
```suggestion
    private AtomicInteger nextTypeAtomIndex = new AtomicInteger(0);
```

Reply:
You may also need to update this test https://github.com/ballerina-platform/ballerina-lang/blob/4982aacc66e9d13696d98cab79d423c83084b5da/semtypes/src/test/java/io/ballerina/types/EnvInitTest.java#L37